File: java/Samples/Biometrics/CS/ServerSampleCS/Controls/EnrollPanel.cs
using System;
using System.Drawing;
using System.Globalization;
using Neurotec.Biometrics;
using Neurotec.Samples.Code;

namespace Neurotec.Samples.Controls
{
	public partial class EnrollPanel : ControlBase
	{
		#region Public constructor

		public EnrollPanel()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		TaskSender _taskSender;

		#endregion

		#region Public methods

		public override string GetTitle()
		{
			return "Enroll Templates";
		}

		public override void Cancel()
		{
			if (!IsBusy) return;

			_taskSender.Cancel();
			btnCancel.Enabled = false;
			AppendStatus("\r\nCanceling, please wait ...\r\n");
		}

		public override bool IsBusy
		{
			get
			{
				if (_taskSender != null)
					return _taskSender.IsBusy;
				return false;
			}
		}

		#endregion

		#region Private form events
		private void BtnStartClick(object sender, EventArgs e)
		{
			try
			{
				if (IsBusy) return;

				pbarProgress.Value = 0;
				int count = GetTemplateCount();
				pbarProgress.Maximum = count;
				tbTaskCount.Text = count.ToString(CultureInfo.InvariantCulture);
				rtbStatus.Text = string.Empty;
				pbStatus.Image = null;
				tbTime.Text = @"N/A";

				_taskSender.TemplateLoader = TemplateLoader;
				_taskSender.IsAccelerator = Accelerator != null;
				_taskSender.BiometricClient = BiometricClient;
				_taskSender.BunchSize = (int)nudBunchSize.Value;
				_taskSender.Start();

				AppendStatus(string.Format("Enrolling from: {0}\r\n", TemplateLoader));
				EnableControls(false);
			}
			catch (Exception ex)
			{
				Utilities.ShowError(ex);
			}
		}

		private void EnrollPanelLoad(object sender, EventArgs e)
		{
			_taskSender = new TaskSender(BiometricClient, TemplateLoader, NBiometricOperations.Enroll);
			_taskSender.ProgressChanged += TaskSenderProgressChanged;
			_taskSender.Finished += TaskSenderFinished;
			_taskSender.ExceptionOccured += TaskSenderExceptionOccured;
			lblProgress.Text = string.Empty;
			lblRemaining.Text = string.Empty;
		}

		private void TaskSenderProgressChanged()
		{
			var templatesEnrolled = _taskSender.PerformedTaskCount;
			if (templatesEnrolled == 1)
			{
				rtbStatus.Text = string.Format("Enrolling from: {0}\r\n", TemplateLoader);
			}
			pbarProgress.Value = templatesEnrolled;

			TimeSpan elapsed = _taskSender.ElapsedTime;
			lblProgress.Text = string.Format("{0} / {1}", templatesEnrolled, tbTaskCount.Text);
			tbTime.Text = elapsed.TotalSeconds.ToString("#.## s");

			TimeSpan remaining = TimeSpan.FromTicks(elapsed.Ticks / templatesEnrolled * (pbarProgress.Maximum - templatesEnrolled));
			if (remaining.TotalSeconds < 0)
				remaining = TimeSpan.Zero;
			lblRemaining.Text = string.Format("Estimated time remaining: {0:00}.{1:00}:{2:00}:{3:00}", remaining.Days, remaining.Hours, remaining.Minutes, remaining.Seconds);

		}

		private void TaskSenderFinished()
		{
			EnableControls(true);
			tbTime.Text = string.Format("{0:00}:{1:00}:{2:00}", _taskSender.ElapsedTime.Hours, _taskSender.ElapsedTime.Minutes, _taskSender.ElapsedTime.Seconds);
			lblRemaining.Text = string.Empty;
			lblProgress.Text = string.Empty;
			if (_taskSender.Canceled)
			{
				AppendStatus("\r\nEnrollment canceled");
				pbStatus.Image = Properties.Resources.error;
				return;
			}

			if (_taskSender.Successful)
			{
				AppendStatus("\r\nAll templates enrolled successfully");
				pbStatus.Image = Properties.Resources.ok;
			}
			else
			{
				AppendStatus("\r\nThere were errors during enrollment", Color.DarkRed);
				pbStatus.Image = Properties.Resources.error;
			}
		}

		private void TaskSenderExceptionOccured(Exception ex)
		{
			AppendStatus(string.Format("{0}\r\n", ex), Color.DarkRed);
		}

		private void BtnCancelClick(object sender, EventArgs e)
		{
			Cancel();
		}

		#endregion

		#region Private methods

		private void EnableControls(bool isIdle)
		{
			btnStart.Enabled = isIdle;
			btnCancel.Enabled = !isIdle;
			nudBunchSize.Enabled = isIdle;
		}

		private void AppendStatus(string msg)
		{
			AppendStatus(msg, Color.Black);
		}

		private delegate void AppendStatusDelegate(string msg, Color color);
		private void AppendStatus(string msg, Color color)
		{
			if (InvokeRequired)
			{
				AppendStatusDelegate del = AppendStatus;
				BeginInvoke(del, msg, color);
				return;
			}

			rtbStatus.SelectionStart = rtbStatus.TextLength;
			rtbStatus.SelectionColor = color;
			rtbStatus.AppendText(msg);
			rtbStatus.SelectionColor = rtbStatus.ForeColor;
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/ServerSampleVB/Controls/EnrollPanel.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Globalization
Imports Neurotec.Biometrics

Partial Public Class EnrollPanel
	Inherits ControlBase
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _taskSender As TaskSender

#End Region

#Region "Public methods"

	Public Overrides Function GetTitle() As String
		Return "Enroll Templates"
	End Function

	Public Overrides Sub Cancel()
		If (Not IsBusy) Then
			Return
		End If

		_taskSender.Cancel()
		btnCancel.Enabled = False
		AppendStatus(Constants.vbCrLf & "Canceling, please wait ..." & Constants.vbCrLf)
	End Sub

	Public Overrides ReadOnly Property IsBusy() As Boolean
		Get
			If _taskSender IsNot Nothing Then
				Return _taskSender.IsBusy
			End If
			Return False
		End Get
	End Property

#End Region

#Region "Private form events"
	Private Sub BtnStartClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStart.Click
		Try
			If IsBusy Then
				Return
			End If

			pbarProgress.Value = 0
			Dim count As Integer = GetTemplateCount()
			pbarProgress.Maximum = count
			tbTaskCount.Text = count.ToString(CultureInfo.InvariantCulture)
			rtbStatus.Text = String.Empty
			pbStatus.Image = Nothing
			tbTime.Text = "N/A"

			_taskSender.TemplateLoader = TemplateLoader
			_taskSender.IsAccelerator = Accelerator IsNot Nothing
			_taskSender.BiometricClient = BiometricClient
			_taskSender.BunchSize = CInt(Fix(nudBunchSize.Value))
			_taskSender.Start()

			AppendStatus(String.Format("Enrolling from: {0}" & Constants.vbCrLf, TemplateLoader))
			EnableControls(False)
		Catch ex As Exception
			Utilities.ShowError(ex)
		End Try
	End Sub

	Private Sub EnrollPanelLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		_taskSender = New TaskSender(BiometricClient, TemplateLoader, NBiometricOperations.Enroll)
		AddHandler _taskSender.ProgressChanged, AddressOf TaskSenderProgressChanged
		AddHandler _taskSender.Finished, AddressOf TaskSenderFinished
		AddHandler _taskSender.ExceptionOccured, AddressOf TaskSenderExceptionOccured
		lblProgress.Text = String.Empty
		lblRemaining.Text = String.Empty
	End Sub

	Private Sub TaskSenderProgressChanged()
		Dim templatesEnrolled = _taskSender.PerformedTaskCount
		If templatesEnrolled = 1 Then
			rtbStatus.Text = String.Format("Enrolling from: {0}" & Constants.vbCrLf, TemplateLoader)
		End If
		pbarProgress.Value = templatesEnrolled

		Dim elapsed As TimeSpan = _taskSender.ElapsedTime
		lblProgress.Text = String.Format("{0} / {1}", templatesEnrolled, tbTaskCount.Text)
		tbTime.Text = elapsed.TotalSeconds.ToString("#.## s")

		Dim remaining As TimeSpan = TimeSpan.FromTicks(elapsed.Ticks / templatesEnrolled * (pbarProgress.Maximum - templatesEnrolled))
		If remaining.TotalSeconds < 0 Then
			remaining = TimeSpan.Zero
		End If
		lblRemaining.Text = String.Format("Estimated time remaining: {0:00}.{1:00}:{2:00}:{3:00}", remaining.Days, remaining.Hours, remaining.Minutes, remaining.Seconds)

	End Sub

	Private Sub TaskSenderFinished()
		EnableControls(True)
		tbTime.Text = String.Format("{0:00}:{1:00}:{2:00}", _taskSender.ElapsedTime.Hours, _taskSender.ElapsedTime.Minutes, _taskSender.ElapsedTime.Seconds)
		lblRemaining.Text = String.Empty
		lblProgress.Text = String.Empty
		If _taskSender.Canceled Then
			AppendStatus(Constants.vbCrLf & "Enrollment canceled")
			pbStatus.Image = My.Resources._error
			Return
		End If

		If _taskSender.Successful Then
			AppendStatus(Constants.vbCrLf & "All templates enrolled successfully")
			pbStatus.Image = My.Resources.ok
		Else
			AppendStatus(Constants.vbCrLf & "There were errors during enrollment", Color.DarkRed)
			pbStatus.Image = My.Resources._error
		End If
	End Sub

	Private Sub TaskSenderExceptionOccured(ByVal ex As Exception)
		AppendStatus(String.Format("{0}" & Constants.vbCrLf, ex), Color.DarkRed)
	End Sub

	Private Sub BtnCancelClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnCancel.Click
		Cancel()
	End Sub

#End Region

#Region "Private methods"

	Private Sub EnableControls(ByVal isIdle As Boolean)
		btnStart.Enabled = isIdle
		btnCancel.Enabled = Not isIdle
		nudBunchSize.Enabled = isIdle
	End Sub

	Private Sub AppendStatus(ByVal msg As String)
		AppendStatus(msg, Color.Black)
	End Sub

	Private Delegate Sub AppendStatusDelegate(ByVal msg As String, ByVal color As Color)
	Private Sub AppendStatus(ByVal msg As String, ByVal color As Color)
		If InvokeRequired Then
			Dim del As AppendStatusDelegate = AddressOf AppendStatus
			BeginInvoke(del, msg, color)
			Return
		End If

		rtbStatus.SelectionStart = rtbStatus.TextLength
		rtbStatus.SelectionColor = color
		rtbStatus.AppendText(msg)
		rtbStatus.SelectionColor = rtbStatus.ForeColor
	End Sub

#End Region
End Class
